data/ledger.go

```diff
@@ -308,15 +308,48 @@ func (l *Ledger) ConsensusParams(r basics.Round) (config.ConsensusParams, error)
 }
 
 // ConsensusVersion gives the consensus version agreed on in a given round,
-// returning an error if we don't have that round or we have an
-// I/O error.
+// returning an error if the consensus version could not be figured using
+// either the block header for the given round, or the latest block header.
 // Implements agreement.Ledger.ConsensusVersion
 func (l *Ledger) ConsensusVersion(r basics.Round) (protocol.ConsensusVersion, error) {
 	blockhdr, err := l.BlockHdr(r)
-	if err != nil {
+	if err == nil {
+		return blockhdr.UpgradeState.CurrentProtocol, nil
+	}
+	// try to see if we can figure out what the version would be.
+	latestRound := l.Latest()
+	// if the request round was for an older round, then just say the we don't know.
+	if r < latestRound {
 		return "", err
 	}
-	return blockhdr.UpgradeState.CurrentProtocol, nil
+	// the request was for a future round. See if we have any known plans for the next round.
+	latestBlockhdr, err := l.BlockHdr(latestRound)
+	// if we have the lastest block header, look inside and try to figure out if we can deduce the
+	// protocol version for the given round.
+	if err == nil {
+		// check to see if we have a protocol upgrade.
+		if latestBlockhdr.NextProtocolSwitchOn == 0 {
+			// no protocol upgrade taking place, which implies the next round has the same version as the current one.
+			if r == (latestBlockhdr.Round + 1) {
+				return latestBlockhdr.CurrentProtocol, nil
+			}
+			// otherwise, we can't really tell.
+			return "", ledgercore.ErrNoEntry{Round: r, Latest: latestRound, Committed: latestRound}
+		}
+		// in this case, we do have a protocol upgrade taking place.
+		if r < latestBlockhdr.NextProtocolSwitchOn {
+			// if we're in the voting duration or uprade waiting period, then the protocol version is the current version.
+			return latestBlockhdr.CurrentProtocol, nil
+		}
+		// if the requested round aligns with the protocol version switch version and we've passed the voting period, then we know that on the switching round
+		// we will be using the next protocol.
+		if r == latestBlockhdr.NextProtocolSwitchOn && latestBlockhdr.Round >= latestBlockhdr.NextProtocolVoteBefore {
```

Comment:
Why latestBlockhdr.Round >= latestBlockhdr.NextProtocolVoteBefore?
Does NextProtocolVoteBefore mean that at that round, there is still voting, which might be voted down?
Shouldn't it be '>' instead of '>='?

Reply:
The `NextProtocolVoteBefore` if I understand correct is the threshold. At that round we already know if the new protocol was adopted or not. ( i.e. since the counting start at 0 )

Reply:
I agree. One more point. Should we also take into account MinUpgradeWaitRounds?